File: ios/Classes/FixitWebviewPlugin.swift
import Flutter
import UIKit
import WebKit
import AVFoundation
import CoreLocation

public class FixitWebviewPlugin: NSObject, FlutterPlugin, FixitWebViewHostApi {
    private var messenger: FlutterBinaryMessenger?
    private var locationManager: CLLocationManager?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let instance = FixitWebviewPlugin()
        instance.messenger = messenger

        // ── Phase A: Memory Pressure Monitoring ──────────────────────────
        NotificationCenter.default.addObserver(
            forName: UIApplication.didReceiveMemoryWarningNotification,
            object: nil,
            queue: .main
        ) { _ in
            FixitWebViewRegistry.broadcastEvent(type: "memoryPressure", value: "critical")
        }

        registrar.register(
            FixitWebViewFactory(messenger: messenger),
            withId: "com.fixit.fixit_webview/view"
        )

        FixitWebViewHostApiSetup.setUp(binaryMessenger: messenger, api: instance)

        let permissionChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/permissions",
            binaryMessenger: messenger
        )
        permissionChannel.setMethodCallHandler(instance.handleMethodCall)

        let uploadChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/upload",
            binaryMessenger: messenger
        )
        uploadChannel.setMethodCallHandler(instance.handleMethodCall)

        let downloadChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/download",
            binaryMessenger: messenger
        )
        downloadChannel.setMethodCallHandler(instance.handleMethodCall)

        let navigationChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/navigation",
            binaryMessenger: messenger
        )
        navigationChannel.setMethodCallHandler(instance.handleMethodCall)

        let offlineChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/offline",
            binaryMessenger: messenger
        )
        offlineChannel.setMethodCallHandler(instance.handleMethodCall)

        let lifecycleChannel = FlutterMethodChannel(
            name: "com.fixit.fixit_webview/lifecycle",
            binaryMessenger: messenger
        )
        lifecycleChannel.setMethodCallHandler(instance.handleMethodCall)
    }

    // MARK: - Method Channel Handler (Permissions + Upload + Download + Navigation + Offline)

    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setCachedResponse":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let url = args["url"] as? String,
                  let data = args["data"] as? String else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            let mimeType = args["mimeType"] as? String ?? "text/html"
            FixitWebViewRegistry.get(viewId: viewId)?.setCachedResponse(url: url, data: data, mimeType: mimeType)
            result(true)
        case "clearOfflineCache":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.clearOfflineCache()
            result(true)
        case "setOfflineFallback":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let html = args["html"] as? String else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.setOfflineFallback(html: html)
            result(true)
        case "checkPermission":
            guard let type = (call.arguments as? [String: Any])?["type"] as? Int else {
                result(1)
                return
            }
            result(checkIosPermission(type: type).rawValue)
        case "requestPermission":
            guard let type = (call.arguments as? [String: Any])?["type"] as? Int else {
                result(1)
                return
            }
            requestIosPermission(type: type, completion: result)
        case "resolveUpload":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let requestId = args["requestId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            let filePaths = args["filePaths"] as? [String] ?? []
            FixitWebViewRegistry.get(viewId: viewId)?.resolveUpload(requestId: requestId, filePaths: filePaths)
            result(true)
        case "cancelUpload":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let requestId = args["requestId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.cancelUpload(requestId: requestId)
            result(true)
        case "startDownload":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let requestId = args["requestId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.startDownload(requestId: requestId)
            result(true)
        case "cancelDownload":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int,
                  let requestId = args["requestId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.cancelDownload(requestId: requestId)
            result(true)
        case "acceptSslError":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.acceptSslError()
            result(true)
        case "denySslError":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.denySslError()
            result(true)
        case "openDownloadedFile":
            guard let args = call.arguments as? [String: Any],
                  let filePath = args["filePath"] as? String else {
                result(false)
                return
            }
            let mimeType = args["mimeType"] as? String ?? ""
            let url = URL(fileURLWithPath: filePath)
            if #available(iOS 9.0, *) {
                let controller = UIDocumentInteractionController(url: url)
                // Present from the root view controller
                if let rootVC = UIApplication.shared.keyWindow?.rootViewController {
                    controller.presentPreview(animated: true)
                }
            }
            result(true)
        case "pauseWebView":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.pause()
            result(true)
        case "resumeWebView":
            guard let args = call.arguments as? [String: Any],
                  let viewIdInt = args["viewId"] as? Int else {
                result(false)
                return
            }
            let viewId = Int64(viewIdInt)
            FixitWebViewRegistry.get(viewId: viewId)?.resume()
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private enum PermissionStatus: Int {
        case granted = 0
        case denied = 1
        case deniedForever = 2
        case restricted = 3
        case limited = 4
    }

    private func checkIosPermission(type: Int) -> PermissionStatus {
        switch type {
        case 0: // camera
            let status = AVCaptureDevice.authorizationStatus(for: .video)
            return mapAvStatus(status)
        case 1: // microphone
            let status = AVCaptureDevice.authorizationStatus(for: .audio)
            return mapAvStatus(status)
        case 2: // location
            let status = CLLocationManager.authorizationStatus()
            return mapLocationStatus(status)
        default:
            return .denied
        }
    }

    private func requestIosPermission(type: Int, completion: @escaping FlutterResult) {
        switch type {
        case 0: // camera
            AVCaptureDevice.requestAccess(for: .video) { granted in
                completion(granted ? PermissionStatus.granted.rawValue : PermissionStatus.denied.rawValue)
            }
        case 1: // microphone
            AVCaptureDevice.requestAccess(for: .audio) { granted in
                completion(granted ? PermissionStatus.granted.rawValue : PermissionStatus.denied.rawValue)
            }
        case 2: // location
            let manager = CLLocationManager()
            self.locationManager = manager
            manager.requestWhenInUseAuthorization()
            // Return current status immediately; delegate callback could be used for live updates
            let status = CLLocationManager.authorizationStatus()
            completion(mapLocationStatus(status).rawValue)
        default:
            completion(PermissionStatus.denied.rawValue)
        }
    }

    private func mapAvStatus(_ status: AVAuthorizationStatus) -> PermissionStatus {
        switch status {
        case .authorized: return .granted
        case .denied: return .denied
        case .restricted: return .restricted
        case .notDetermined: return .denied
        @unknown default: return .denied
        }
    }

    private func mapLocationStatus(_ status: CLAuthorizationStatus) -> PermissionStatus {
        switch status {
        case .authorizedWhenInUse, .authorizedAlways: return .granted
        case .denied: return .denied
        case .restricted: return .restricted
        case .notDetermined: return .denied
        @unknown default: return .denied
        }
    }

    // MARK: - FixitWebViewHostApi

    public func create(viewId: Int64, config: PigeonRuntimeConfig) throws {
        if let webView = FixitWebViewRegistry.get(viewId: viewId) {
            webView.applyConfig(config: config)
        } else {
            FixitWebViewRegistry.setPendingConfig(viewId: viewId, config: config)
        }
    }

    public func loadUrl(viewId: Int64, url: String) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.loadUrl(url: url)
    }

    public func loadUrlWithHeaders(viewId: Int64, url: String, headers: [String?: String?]?, method: String?, body: String?) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.loadUrlWithHeaders(url: url, headers: headers, method: method, body: body)
    }

    public func loadHtmlString(viewId: Int64, html: String, baseUrl: String?) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.loadHtmlString(html: html, baseUrl: baseUrl)
    }

    public func stopLoading(viewId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.stopLoading()
    }

    public func getTitle(viewId: Int64) throws -> String? {
        return FixitWebViewRegistry.get(viewId: viewId)?.getTitle()
    }

    public func goBack(viewId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.goBack()
    }

    public func goForward(viewId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.goForward()
    }

    public func reload(viewId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.reload()
    }

    public func clearCache(viewId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.clearCache()
    }

    public func clearCookies() throws {
        let cookieStore = WKWebsiteDataStore.default().httpCookieStore
        cookieStore.getAllCookies { cookies in
            for cookie in cookies {
                cookieStore.delete(cookie, completionHandler: nil)
            }
        }
    }

    public func setCookie(url: String, key: String, value: String) throws {
        if let nsUrl = URL(string: url),
           let cookie = HTTPCookie(properties: [
            .domain: nsUrl.host ?? "",
            .path: nsUrl.path,
            .name: key,
            .value: value,
            .secure: nsUrl.scheme == "https" ? "TRUE" : "FALSE"
           ]) {
            WKWebsiteDataStore.default().httpCookieStore.setCookie(cookie, completionHandler: nil)
        }
    }

    public func getCookies(url: String, completion: @escaping (Result<[String?], Error>) -> Void) {
        guard let nsUrl = URL(string: url) else {
            completion(.success([]))
            return
        }
        let cookieStore = WKWebsiteDataStore.default().httpCookieStore
        cookieStore.getAllCookies { cookies in
            let matching = cookies
                .filter { $0.domain == nsUrl.host || nsUrl.host?.hasSuffix($0.domain) == true }
                .map { "\($0.name)=\($0.value)" }
            completion(.success(matching))
        }
    }

    public func postBridgeMessage(viewId: Int64, message: String) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.postBridgeMessage(message: message)
    }

    public func httpAuthResponse(viewId: Int64, requestId: Int64, username: String, password: String) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.httpAuthResponse(requestId: Int(requestId), username: username, password: password)
    }

    public func cancelHttpAuth(viewId: Int64, requestId: Int64) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.cancelHttpAuth(requestId: Int(requestId))
    }

    public func updateSecurityConfig(viewId: Int64, mixedContentMode: Int64?, safeBrowsingEnabled: Bool?, zoomEnabled: Bool?) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.updateSecurityConfig(mixedContentMode: mixedContentMode, safeBrowsingEnabled: safeBrowsingEnabled, zoomEnabled: zoomEnabled)
    }

    public func evaluateJavascript(viewId: Int64, javascript: String) throws {
        FixitWebViewRegistry.get(viewId: viewId)?.evaluateJavascript(javascript: javascript)
    }

    public func runJavascriptReturningResult(viewId: Int64, javascript: String, completion: @escaping (Result<String?, Error>) -> Void) {
        if let view = FixitWebViewRegistry.get(viewId: viewId) {
            view.runJavascriptReturningResult(javascript: javascript, completion: completion)
        } else {
            completion(.failure(NSError(domain: "FixitWebviewPlugin", code: 404, userInfo: [NSLocalizedDescriptionKey: "WebView not found"])))
        }
    }

    public func dispose(viewId: Int64) throws {
        FixitWebViewRegistry.remove(viewId: viewId)?.dispose()
    }
}

File: android/src/main/kotlin/com/fixit/fixit_webview/FixitWebviewPlugin.kt
package com.fixit.fixit_webview

import android.Manifest
import android.content.ComponentCallbacks2
import android.content.res.Configuration
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.webkit.CookieManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.fixit.fixit_webview.webview.FixitWebViewFactory
import com.fixit.fixit_webview.webview.FixitWebViewHostApi
import com.fixit.fixit_webview.webview.FixitWebViewRegistry
import com.fixit.fixit_webview.webview.PigeonRuntimeConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

/**
 * Main Flutter plugin entry point.
 * Registers the PlatformView factory, the Pigeon host API, and the permission MethodChannel.
 */
class FixitWebviewPlugin : FlutterPlugin, FixitWebViewHostApi, ActivityAware, MethodChannel.MethodCallHandler, ComponentCallbacks2 {

    private var context: Context? = null
    private var application: android.app.Application? = null
    private var messenger: BinaryMessenger? = null
    private var activity: android.app.Activity? = null
    private var permissionChannel: MethodChannel? = null
    private var uploadChannel: MethodChannel? = null
    private var downloadChannel: MethodChannel? = null
    private var navigationChannel: MethodChannel? = null
    private var offlineChannel: MethodChannel? = null
    private var lifecycleChannel: MethodChannel? = null

    // ── Permission request tracking ────────────────────────────────
    private val pendingPermissionRequests = mutableMapOf<Int, MethodChannel.Result>()
    private var permissionRequestCode = 3000
    private val permissionRequestCodeMap = mutableMapOf<Int, List<String>>()

    companion object {
        private const val PERMISSION_CHANNEL = "com.fixit.fixit_webview/permissions"
        private const val UPLOAD_CHANNEL = "com.fixit.fixit_webview/upload"
        private const val DOWNLOAD_CHANNEL = "com.fixit.fixit_webview/download"
        private const val NAVIGATION_CHANNEL = "com.fixit.fixit_webview/navigation"
        private const val OFFLINE_CHANNEL = "com.fixit.fixit_webview/offline"
        private const val LIFECYCLE_CHANNEL = "com.fixit.fixit_webview/lifecycle"
    }

    // ─── FlutterPlugin ────────────────────────────────────────
    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        messenger = binding.binaryMessenger

        // Register for memory pressure events
        if (binding.applicationContext is android.app.Application) {
            application = binding.applicationContext as android.app.Application
            application?.registerComponentCallbacks(this)
        }

        binding.platformViewRegistry.registerViewFactory(
            "com.fixit.fixit_webview/view",
            FixitWebViewFactory(binding.binaryMessenger)
        )

        FixitWebViewHostApi.setUp(binding.binaryMessenger, this)

        permissionChannel = MethodChannel(binding.binaryMessenger, PERMISSION_CHANNEL)
        permissionChannel?.setMethodCallHandler(this)
        uploadChannel = MethodChannel(binding.binaryMessenger, UPLOAD_CHANNEL)
        uploadChannel?.setMethodCallHandler(this)
        downloadChannel = MethodChannel(binding.binaryMessenger, DOWNLOAD_CHANNEL)
        downloadChannel?.setMethodCallHandler(this)
        navigationChannel = MethodChannel(binding.binaryMessenger, NAVIGATION_CHANNEL)
        navigationChannel?.setMethodCallHandler(this)
        offlineChannel = MethodChannel(binding.binaryMessenger, OFFLINE_CHANNEL)
        offlineChannel?.setMethodCallHandler(this)
        lifecycleChannel = MethodChannel(binding.binaryMessenger, LIFECYCLE_CHANNEL)
        lifecycleChannel?.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        application?.unregisterComponentCallbacks(this)
        application = null
        FixitWebViewHostApi.setUp(binding.binaryMessenger, null)
        permissionChannel?.setMethodCallHandler(null)
        permissionChannel = null
        uploadChannel?.setMethodCallHandler(null)
        uploadChannel = null
        downloadChannel?.setMethodCallHandler(null)
        downloadChannel = null
        navigationChannel?.setMethodCallHandler(null)
        navigationChannel = null
        offlineChannel?.setMethodCallHandler(null)
        offlineChannel = null
        lifecycleChannel?.setMethodCallHandler(null)
        lifecycleChannel = null
        context = null
        messenger = null
    }

    // ─── ActivityAware ────────────────────────────────────────────
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(permissionResultListener)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // ── Phase A: Memory Pressure (ComponentCallbacks2) ────────────────────────
    override fun onTrimMemory(level: Int) {
        val pressureLevel = when {
            level >= android.content.ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL -> "critical"
            level >= android.content.ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE -> "moderate"
            else -> "none"
        }
        if (pressureLevel != "none") {
            FixitWebViewRegistry.broadcastEvent("memoryPressure", pressureLevel)
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {}
    override fun onLowMemory() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    // ─── Permission Result Listener ───────────────────────────────
    private val permissionResultListener = PluginRegistry.RequestPermissionsResultListener { requestCode, permissions, grantResults ->
        val code = requestCode
        if (permissionRequestCodeMap.containsKey(code)) {
            val requestedPermissions = permissionRequestCodeMap.remove(code)
            val result = pendingPermissionRequests.remove(code)
            if (result != null && requestedPermissions != null) {
                val allGranted = grantResults.all { it == PackageManager.PERMISSION_GRANTED }
                result.success(if (allGranted) 0 else 1) // 0 = granted, 1 = denied
            }
            true
        } else {
            false
        }
    }

    // ─── MethodChannel (Permissions) ──────────────────────────────
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val act = activity
        val ctx = context
        if (act == null || ctx == null) {
            result.error("NO_ACTIVITY", "Plugin not attached to activity", null)
            return
        }

        when (call.method) {
            "startDownload" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val requestId = call.argument<Int>("requestId") ?: run {
                    result.success(false)
                    return
                }
                val destinationDir = call.argument<String>("destinationDir")
                val view = FixitWebViewRegistry.get(viewId)
                if (view != null) {
                    // The native listener already has the URL. We start the download using
                    // the DownloadManager. Since we don't store the URL per requestId on the native
                    // side (it was already sent via event), we use the view's generic startDownload
                    // that fetches from the pending event context.
                    view.startDownload(requestId, destinationDir)
                }
                result.success(true)
            }
            "cancelDownload" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val requestId = call.argument<Int>("requestId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.cancelDownload(requestId)
                result.success(true)
            }
            "resolveUpload" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val requestId = call.argument<Int>("requestId") ?: run {
                    result.success(false)
                    return
                }
                val filePaths = call.argument<List<String>>("filePaths") ?: emptyList()
                FixitWebViewRegistry.get(viewId)?.resolveUpload(requestId, filePaths)
                result.success(true)
            }
            "cancelUpload" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val requestId = call.argument<Int>("requestId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.cancelUpload(requestId)
                result.success(true)
            }
            "acceptSslError" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.acceptSslError()
                result.success(true)
            }
            "denySslError" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.denySslError()
                result.success(true)
            }
            "setCachedResponse" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val url = call.argument<String>("url") ?: run {
                    result.success(false)
                    return
                }
                val data = call.argument<String>("data") ?: run {
                    result.success(false)
                    return
                }
                val mimeType = call.argument<String>("mimeType") ?: "text/html"
                FixitWebViewRegistry.get(viewId)?.setCachedResponse(url, data, mimeType)
                result.success(true)
            }
            "clearOfflineCache" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.clearOfflineCache()
                result.success(true)
            }
            "setOfflineFallback" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                val html = call.argument<String>("html") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.setOfflineFallback(html)
                result.success(true)
            }
            "openDownloadedFile" -> {
                val filePath = call.argument<String>("filePath") ?: run {
                    result.success(false)
                    return
                }
                val mimeType = call.argument<String>("mimeType") ?: run {
                    result.success(false)
                    return
                }
                try {
                    val uri = Uri.parse(filePath)
                    val intent = Intent(Intent.ACTION_VIEW).apply {
                        setDataAndType(uri, mimeType)
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    }
                    act.startActivity(intent)
                    result.success(true)
                } catch (e: Exception) {
                    result.success(false)
                }
            }
            "checkPermission" -> {
                val type = call.argument<Int>("type") ?: 0
                val permission = permissionForType(type)
                if (permission == null) {
                    result.success(1) // denied
                    return
                }
                val status = checkAndroidPermission(ctx, permission)
                result.success(status.ordinal)
            }
            "requestPermission" -> {
                val type = call.argument<Int>("type") ?: 0
                val permission = permissionForType(type)
                if (permission == null) {
                    result.success(1) // denied
                    return
                }
                if (checkAndroidPermission(ctx, permission) == PermissionStatus.GRANTED) {
                    result.success(0) // granted
                    return
                }
                val code = permissionRequestCode++
                pendingPermissionRequests[code] = result
                permissionRequestCodeMap[code] = listOf(permission)
                ActivityCompat.requestPermissions(act, arrayOf(permission), code)
            }
            "pauseWebView" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.pause()
                result.success(true)
            }
            "resumeWebView" -> {
                val viewId = call.argument<Int>("viewId") ?: run {
                    result.success(false)
                    return
                }
                FixitWebViewRegistry.get(viewId)?.resume()
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    private enum class PermissionStatus {
        GRANTED, DENIED
    }

    private fun checkAndroidPermission(ctx: Context, permission: String): PermissionStatus {
        return if (ContextCompat.checkSelfPermission(ctx, permission) == PackageManager.PERMISSION_GRANTED) {
            PermissionStatus.GRANTED
        } else {
            PermissionStatus.DENIED
        }
    }

    private fun permissionForType(type: Int): String? {
        return when (type) {
            0 -> Manifest.permission.CAMERA
            1 -> Manifest.permission.RECORD_AUDIO
            2 -> Manifest.permission.ACCESS_FINE_LOCATION
            else -> null
        }
    }

    // ─── FixitWebViewHostApi (Pigeon) ─────────────────────────
    override fun create(viewId: Long, config: PigeonRuntimeConfig) {
        val id = viewId.toInt()
        val view = FixitWebViewRegistry.get(id)
        if (view != null) {
            view.applyConfig(config)
        } else {
            FixitWebViewRegistry.setPendingConfig(id, config)
        }
    }

    override fun loadUrl(viewId: Long, url: String) {
        FixitWebViewRegistry.get(viewId.toInt())?.loadUrl(url)
    }

    override fun loadUrlWithHeaders(viewId: Long, url: String, headers: Map<String?, String?>?, method: String?, body: String?) {
        FixitWebViewRegistry.get(viewId.toInt())?.loadUrlWithHeaders(url, headers, method, body)
    }

    override fun loadHtmlString(viewId: Long, html: String, baseUrl: String?) {
        FixitWebViewRegistry.get(viewId.toInt())?.loadHtmlString(html, baseUrl)
    }

    override fun stopLoading(viewId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.stopLoading()
    }

    override fun getTitle(viewId: Long): String? {
        return FixitWebViewRegistry.get(viewId.toInt())?.getTitle()
    }

    override fun goBack(viewId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.goBack()
    }

    override fun goForward(viewId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.goForward()
    }

    override fun reload(viewId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.reload()
    }

    override fun clearCache(viewId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.clearCache()
    }

    override fun clearCookies() {
        val cookieManager = CookieManager.getInstance()
        cookieManager.removeAllCookies(null)
        cookieManager.flush()
    }

    override fun setCookie(url: String, key: String, value: String) {
        val cookieManager = CookieManager.getInstance()
        cookieManager.setAcceptCookie(true)
        cookieManager.setCookie(url, "$key=$value")
    }

    override fun getCookies(url: String, callback: (Result<List<String?>>) -> Unit) {
        val raw = CookieManager.getInstance().getCookie(url) ?: ""
        val cookies = if (raw.isEmpty()) emptyList() else raw.split(";").map { it.trim() }
        callback(Result.success(cookies))
    }

    override fun postBridgeMessage(viewId: Long, message: String) {
        FixitWebViewRegistry.get(viewId.toInt())?.postBridgeMessage(message)
    }

    override fun httpAuthResponse(viewId: Long, requestId: Long, username: String, password: String) {
        FixitWebViewRegistry.get(viewId.toInt())?.httpAuthResponse(requestId.toInt(), username, password)
    }

    override fun cancelHttpAuth(viewId: Long, requestId: Long) {
        FixitWebViewRegistry.get(viewId.toInt())?.cancelHttpAuth(requestId.toInt())
    }

    override fun updateSecurityConfig(viewId: Long, mixedContentMode: Long?, safeBrowsingEnabled: Boolean?, zoomEnabled: Boolean?) {
        FixitWebViewRegistry.get(viewId.toInt())?.updateSecurityConfig(mixedContentMode, safeBrowsingEnabled, zoomEnabled)
    }

    override fun evaluateJavascript(viewId: Long, javascript: String) {
        FixitWebViewRegistry.get(viewId.toInt())?.evaluateJavascript(javascript)
    }

    override fun runJavascriptReturningResult(
        viewId: Long,
        javascript: String,
        callback: (Result<String?>) -> Unit
    ) {
        val view = FixitWebViewRegistry.get(viewId.toInt())
        if (view != null) {
            view.runJavascriptReturningResult(javascript, callback)
        } else {
            callback(Result.failure(Exception("WebView not found for viewId: $viewId")))
        }
    }

    override fun dispose(viewId: Long) {
        FixitWebViewRegistry.remove(viewId.toInt())?.dispose()
    }
}
